zeppelin-zengine/src/main/java/org/apache/zeppelin/interpreter/launcher/ShellScriptLauncher.java

```diff
@@ -21,50 +21,67 @@
 import org.apache.zeppelin.conf.ZeppelinConfiguration;
 import org.apache.zeppelin.interpreter.InterpreterOption;
 import org.apache.zeppelin.interpreter.InterpreterRunner;
+import org.apache.zeppelin.interpreter.recovery.RecoveryStorage;
 import org.apache.zeppelin.interpreter.remote.RemoteInterpreterManagedProcess;
 import org.apache.zeppelin.interpreter.remote.RemoteInterpreterRunningProcess;
 import org.apache.zeppelin.interpreter.remote.RemoteInterpreterUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
 import java.util.HashMap;
 import java.util.Map;
 
 /**
  * Interpreter Launcher which use shell script to launch the interpreter process.
- *
  */
 public class ShellScriptLauncher extends InterpreterLauncher {
 
   private static final Logger LOGGER = LoggerFactory.getLogger(ShellScriptLauncher.class);
 
-  public ShellScriptLauncher(ZeppelinConfiguration zConf) {
-    super(zConf);
+  public ShellScriptLauncher(ZeppelinConfiguration zConf, RecoveryStorage recoveryStorage) {
+    super(zConf, recoveryStorage);
   }
 
   @Override
-  public InterpreterClient launch(InterpreterLaunchContext context) {
+  public InterpreterClient launch(InterpreterLaunchContext context) throws IOException {
     LOGGER.info("Launching Interpreter: " + context.getInterpreterSettingGroup());
     this.properties = context.getProperties();
     InterpreterOption option = context.getOption();
     InterpreterRunner runner = context.getRunner();
     String groupName = context.getInterpreterSettingGroup();
     String name = context.getInterpreterSettingName();
-
     int connectTimeout =
         zConf.getInt(ZeppelinConfiguration.ConfVars.ZEPPELIN_INTERPRETER_CONNECT_TIMEOUT);
+
     if (option.isExistingProcess()) {
       return new RemoteInterpreterRunningProcess(
+          context.getInterpreterSettingName(),
           connectTimeout,
           option.getHost(),
           option.getPort());
     } else {
+      // try to recover it first
+      if (zConf.isRecoveryEnabled()) {
+        Map<String, InterpreterClient> clients = recoveryStorage.restore();
+        if (clients.containsKey(context.getInterpreterGroupId())) {
+          InterpreterClient client = clients.get(context.getInterpreterGroupId());
+          if (client.isRunning()) {
+            LOGGER.info("Recover InterpreterProcess: " + client.getHost() + ":" + client.getPort());
+            return (RemoteInterpreterRunningProcess) client;
+          } else {
+            LOGGER.warn("Can not recovery interpreter process: " + client.getHost() + ":"
```

Comment:
"Cannot recover ..."

Reply:
so when it cannot recover, should it delete the recovery data or should it try again the next time?

Reply:
For `FileSystemRecoveryStorage`, it would refresh all the recovery data when interpreter process is started or stopped. So when recovery is failed, new interpreter process is created and recovery data will be refreshed,  the old recovery data will be deleted as well.